**Student** &nbsp;: *Steven Rankine* <br>
**Subject** &nbsp;: *Data Analytics Boot Camp, Fall 2018* <br>
**Homework** : *Week 2 - In Class Exercises* <br>
**Date** &nbsp;: *2018-10-13* <br>
- - -

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import glob

# Algebra Grades

Frank Mulligrew is the Algebra Coordinator for DC Public Schools. He is required to gather some statistics by the School Board. Using the following information about his class, calculate the following:

1. Percentage of students with a passing grade
2. Percentage of women with a passing grade
3. Average hours of study for all students
4. Average hours of study for students with a passing grade

In [2]:
# Bring dataset into the python environment
df = pd.read_csv("../datasets/algebradata.csv")
df.head()

Fname     Lname Gender Grade  Hours of Study
0     Mary  Ettienne      F     B              16
1  Charles    Looner      M     F               8
2    Betty  Franklin      F     A              24
3    Roger   Withers      M     C               5
4     John   Mulgrew      M     A               5

### *Solution:*
Based on the assumption that grades of A, B, and C are considered passing. Any other letter grade is considered failing.

In [3]:
# Calculate percentage of student passing
test1 = (df['Grade']=='A') | (df['Grade']=='B') | (df['Grade']=='C')
pct_pass = df.loc[test1].count().Grade / df.count().Grade
print('\n%d. Percentage of students with a passing grade:              %6.2f%%'% (1, 100*pct_pass))

# Calculate percentage of women passing
test2 = test1 & (df['Gender']=='F')
pct_pass_f = df.loc[test2].count().Grade / df.count().Grade
print('%d. Percentage of women with a passing grade:                 %6.2f%%'  % (2, 100*pct_pass_f))

# Calculate Average study hours of all students
avg_hours = df['Hours of Study'].mean()
print('%d. Average hours of study for all students:                  %7.3f hrs'% (3, avg_hours))

# Calculate Average study hours of passing students
avg_hours_pass = df.loc[test1].mean()
print('%d. Average hours of study for students with a passing grade: %7.3f hrs'% (4, avg_hours_pass))


1. Percentage of students with a passing grade:               69.17%
2. Percentage of women with a passing grade:                  38.14%
3. Average hours of study for all students:                   15.115 hrs
4. Average hours of study for students with a passing grade:  16.298 hrs


# Axis Auto Sales

Carlos Hugens is the Sales Manager for Axis Auto Sales a low cost regional chain of Used Car Lots.
Carlos is getting ready for his annual sales meeting and he is looking for the best way to improve his
sales group's performance.
His data includes the gender, years of experience, sales training and hours worked per week. It also
includes the average cars sold per month by each salesperson.
Find out the following:

1. Average cars sold per month
2. Max cars sold per month
3. Min cars sold per month
4. Average cars sold per month by gender
5. Average hours worked by people selling > 3 cars per month
6. Average years of experience
7. Average years of experience for people selling > 3 cars per month
8. Average cars sold per month by whether or not they have had sales training
9. What do you think is the best indicator of whether someone is a good salesperson?

- - -
### *Solution:*

In [4]:
# Bring dataset into the python environment
df = pd.read_csv("../datasets/axisdata.csv") # overwite previuos df
df.head()                                    # Show first 5 rows

Fname      Lname Gender  Hours Worked SalesTraining  Years Experience  \
0     Jada   Walters       F            39             N                 3   
1   Nicole  Henderson      F            46             N                 3   
2    Tanya      Moore      F            42             Y                 4   
3  Ronelle    Jackson      F            38             Y                 5   
4     Brad      Sears      M            33             N                 4   

   Cars Sold  
0          2  
1          6  
2          6  
3          3  
4          2

The first three questions are about descriptive statistics of the response variable/feature (*Cars Sold*) for the entire population.

In [5]:
# Calculate statitics of average cars sold per month 
allMin = df['Cars Sold'].min()  # Minimum 
allMax = df['Cars Sold'].max()  # Maximum 
allAvg = df['Cars Sold'].mean() # Mean 
allStd = df['Cars Sold'].std()  # Std Dev 
allVar = df['Cars Sold'].var()  # Variance 

# Display Results
print('%d. Average cars sold per month:  %.3f' % (1, df['Cars Sold'].mean()))
print('%d. Max cars sold per month:      %d' % (2, df['Cars Sold'].max()))
print('%d. Min cars sold per month:      %d' % (3, df['Cars Sold'].min()))

1. Average cars sold per month:  3.923
2. Max cars sold per month:      7
3. Min cars sold per month:      1


The next question is about descriptive statistics of the response variable/feature (*Cars Sold*) for subsets of the population: male-only & female-only.

In [6]:
# Calculate averages based on gender
mAvg = df.loc[df['Gender']=='F','Cars Sold'].mean()
mVar = df.loc[df['Gender']=='F','Cars Sold'].var()
fAvg = df.loc[df['Gender']=='M','Cars Sold'].mean()
fVar = df.loc[df['Gender']=='M','Cars Sold'].var()

# Display Results
print('%d. Average cars sold per month by gender:  %.3f (Female), %.3f (Male)' % (4,mAvg,fAvg))

4. Average cars sold per month by gender:  3.761 (Female), 4.078 (Male)


The next four questions 

In [8]:
# Additional Calculations
hwAvg    = df.loc[df['Cars Sold']>3,'Hours Worked'].mean()
yeAvg    = df['Years Experience'].mean()
yeVar    = df['Years Experience'].var()
yeAvgCS3 = df.loc[df['Cars Sold']>3,'Years Experience'].mean()   
tYes     = df.loc[df['SalesTraining']=='Y', 'Cars Sold'].mean()
tYesVar  = df.loc[df['SalesTraining']=='Y', 'Cars Sold'].var()
tNo      = df.loc[df['SalesTraining']=='N', 'Cars Sold'].mean()
tNoVar   = df.loc[df['SalesTraining']=='N', 'Cars Sold'].var()

# Display Results
print('%d. Average hours worked by people selling > 3 cars per month:  %.2f hrs'% (5, hwAvg))
print('%d. Average years of experience:                                %.3f yrs'% (6, yeAvg))
print('%d. Average years of experience for people selling > 3 cars:    %.3f yrs'% (7,yeAvgCS3))
print('%d. Average cars sold per month based on having sales training: %.3f (Yes), %.3f (No)'% (8,tYes,tNo))

5. Average hours worked by people selling > 3 cars per month:  34.74 hrs
6. Average years of experience:                                3.026 yrs
7. Average years of experience for people selling > 3 cars:    3.049 yrs
8. Average cars sold per month based on having sales training: 4.169 (Yes), 3.574 (No)


Choosing the feature that is te best indicator of sales performance would require the a statistical test like ANOVA which would measure the contribution of each feature (Gender, Hours Worked, SalesTraining, Years Experience) to the overall variance of Cars Sold among the population of sales persons.

In [16]:
m_tYes     = df.loc[(df['SalesTraining']=='Y') & (df['Gender']=='M'), 'Cars Sold'].mean()
m_tYesVar  = df.loc[(df['SalesTraining']=='Y') & (df['Gender']=='M'), 'Cars Sold'].var()
m_tNo      = df.loc[(df['SalesTraining']=='N') & (df['Gender']=='M'), 'Cars Sold'].mean()
m_tNoVar   = df.loc[(df['SalesTraining']=='N') & (df['Gender']=='M'), 'Cars Sold'].var()
f_tYes     = df.loc[(df['SalesTraining']=='Y') & (df['Gender']=='F'), 'Cars Sold'].mean()
f_tYesVar  = df.loc[(df['SalesTraining']=='Y') & (df['Gender']=='F'), 'Cars Sold'].var()
f_tNo      = df.loc[(df['SalesTraining']=='N') & (df['Gender']=='F'), 'Cars Sold'].mean()
f_tNoVar   = df.loc[(df['SalesTraining']=='N') & (df['Gender']=='F'), 'Cars Sold'].var()
yeAvgAbove = df.loc[df['Years Experience']>= yeAvg,'Cars Sold'].mean()   
yeAvgBelow = df.loc[df['Years Experience']<  yeAvg,'Cars Sold'].mean()   
yeVarAbove = df.loc[df['Years Experience']>= yeAvg,'Cars Sold'].var()   
yeVarBelow = df.loc[df['Years Experience']<  yeAvg,'Cars Sold'].var() 

### **Further examination**
There was also a significant delta between male and female average monthly sales figures. If gender were a factor in monthly sales then males with sales training would have even higher average monthly sales.

Factors          |   Avg |   Var |  Delta
:---             |  ---: |  ---: |   ---:
All              | 3.923 | 2.332 |  0.000
Male             | 4.078 | 2.199 | +0.155
Female           | 3.761 | 2.414 | -0.162
Trained          | 4.169 | 2.192 | +0.246
Untrained        | 3.574 | 2.328 | -0.349
Untrained Female | 3.514 | 2.166 | -0.409
Untrained Male   | 3.637 | 2.502 | -0.286
Trained Female   | 3.949 | 2.150 | +0.026
Trained Male     | 4.366 | 2.155 | +0.443
Above Avg Exp    | 4.001 | 2.369 | +0.082
Below Avg Exp    | 3.869 | 2.304 | -0.054

In [17]:

print('9. What is the best indicator of someone being a good salesperson:  Sales Training' )


9. What is the best indicator of someone being a good salesperson:  Sales Training
